### PR TITLE
Log contextualisable subsystems

### PR DESCRIPTION
**Pull request by clero**
Improvements of #58 , Subsystems are now able to use Context when Logging (and the same API as the ParameterMgr)

### CODE REVIEW

**Comment by clero** [comment]
C bindings coverage regression need to be fixed

File: parameter/ParameterMgr.h [hunk 1]
```diff
@@ -624,9 +635,8 @@ class CParameterMgr : private CElement
     // Blackboard access mutex
     pthread_mutex_t _blackboardMutex;
 
-    // Logging
-    ILogger* _pLogger;
-    mutable uint32_t _uiLogDepth;
+    /** */
```
Comment (clero) [inline parameter/ParameterMgr.h]: Missing doc

File: parameter/log/src/Context.cpp [hunk 2]
```diff
@@ -27,27 +27,26 @@
...
-    if (_bLogOn) {
-        // Unnest
-        _pElement->unnestLog();
-        // Log
-        _pElement->doLog(false, "} " + _strContext);
-    }
+    mLogger.mProlog.resize(mLogger.mProlog.size() - 4);
+    mLogger.info() << "}";
 }
+
```
Comment (clero) [inline parameter/log/src/Context.cpp]: Move the entire file to header ?

File: parameter/ParameterMgr.h (hunk 1 above)
Comment (clero) [inline parameter/ParameterMgr.h]: done

File: parameter/log/src/Context.cpp (hunk 2 above)
Comment (clero) [inline parameter/log/src/Context.cpp]: done

File: parameter/ParameterMgr.cpp [hunk 3]
```diff
@@ -492,14 +462,14 @@ bool CParameterMgr::load(string& strError)
...
     // Log selection criterion states
     {
-        CAutoLog autoLog(this, "Criterion states");
+        LOG_CONTEXT("Criterion states");
 
         const CSelectionCriteria* selectionCriteria = getConstSelectionCriteria();
 
-        list<string> lstrSelectionCriteron;
-        selectionCriteria->listSelectionCriteria(lstrSelectionCriteron, true, false);
+        core::Results criteria;
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: This is an opportunistic type reuse.

File: parameter/ParameterMgr.cpp [hunk 4]
```diff
@@ -538,17 +508,40 @@ bool CParameterMgr::loadFrameworkConfiguration(string& strError)
...
+
+    // Load subsystems
+    bool isSuccess = getSystemClass()->loadSubsystems(error,
+                                                      _pSubsystemPlugins,
+                                                      !_bFailOnMissingSubsystem);
+
+    if (isSuccess) {
+        _logger.info() << "All subsystem plugins successfully loaded";
+
+        if(!_bFailOnMissingSubsystem) {
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: `!error.empty() &&`

File: parameter/ParameterMgr.cpp [hunk 5]
```diff
@@ -592,8 +587,8 @@ bool CParameterMgr::loadSettings(string& strError)
 
     if (!success && !_bFailOnFailedSettingsLoad) {
         // Load can not fail, ie continue but log the load errors
-        log_info(strLoadError);
-        log_info("Failed to load settings, continue without domains.");
+        _logger.info() << strLoadError;
+        _logger.info() << "Failed to load settings, continue without domains.";
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: _logger.info() and _logger.warning() could be wrapped in a private method in ParameterMgr

File: parameter/ParameterMgr.cpp [hunk 6]
```diff
@@ -945,11 +945,11 @@ CParameterMgr::CCommandHandler::CommandStatus CParameterMgr::statusCommandProces
 
     /// Criteria states
     appendTitle(strResult, "Selection Criteria:");
-    list<string> lstrSelectionCriteria;
-    getSelectionCriteria()->listSelectionCriteria(lstrSelectionCriteria, false, true);
+    core::Results criteria;
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: same as above (opportunistic)

File: parameter/ParameterMgr.cpp [hunk 7]
```diff
@@ -1132,11 +1132,11 @@ CParameterMgr::CCommandHandler::CommandStatus CParameterMgr::listCriteriaCommman
         // Requested format will be either CSV or human readable based on strOutputFormat content
         bool bHumanReadable = strOutputFormat.empty();
 
-        list<string> lstrResult;
-        getSelectionCriteria()->listSelectionCriteria(lstrResult, true, bHumanReadable);
+        core::Results result;
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: ditto

File: parameter/ParameterMgr.cpp [hunk 8]
```diff
@@ -2632,3 +2724,16 @@ bool CParameterMgr::exportElementToXMLString(const IXmlSource* pXmlSource,
 
     return bProcessSuccess;
 }
+
+bool CParameterMgr::logResult(bool isSuccess, const std::string& result)
+{
+    std::string log = (result.empty() ? "" : ": " + result);
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: useless parentheses

File: parameter/ParameterMgr.h [hunk 9]
```diff
@@ -253,8 +245,17 @@ class CParameterMgr : private CElement
     bool deleteConfiguration(const std::string& strDomain, const std::string& strConfiguration, std::string& strError);
     bool renameConfiguration(const std::string& strDomain, const std::string& strConfiguration, const std::string& strNewConfiguration, std::string& strError);
 
-    // Save/Restore
-    bool restoreConfiguration(const std::string& strDomain, const std::string& strConfiguration, std::list<std::string>& strError);
+    /** Restore a configuration
+     *
+     * @param[in] strDomain the domain name
+     * @param[in] strConfiguration the configuration name
+     * @param[out] errors, errors encountered during restoration
```
Comment (dawagner) [inline parameter/ParameterMgr.h]: superfluous comma

File: parameter/Results.h [hunk 10]
```diff
@@ -27,27 +27,15 @@
...
-{
-    if (_bLogOn) {
-        // Unnest
-        _pElement->unnestLog();
-        // Log
-        _pElement->doLog(false, "} " + _strContext);
-    }
-}
+    /** String list type which can hold list of error/info */
+    typedef std::list<std::string> Results ;
```
Comment (dawagner) [inline parameter/Results.h]: superfluous space before `;`.

File: parameter/SystemClass.cpp [hunk 11]
```diff
@@ -98,38 +101,29 @@ bool CSystemClass::loadSubsystems(string& strError,
...
-    CAutoLog autoLog_info(this, "Loading subsystem plugins");
-
     // Start clean
     _pSubsystemLibrary->clean();
 
+    typedef TLoggingElementBuilderTemplate<CVirtualSubsystem> VirtualSubsystemBuilder;
     // Add virtual subsystem builder
-    _pSubsystemLibrary->addElementBuilder("Virtual",
-                                          new TNamedElementBuilderTemplate<CVirtualSubsystem>());
+    _pSubsystemLibrary->addElementBuilder( "Virtual", new VirtualSubsystemBuilder(_logger));
```
Comment (dawagner) [inline parameter/SystemClass.cpp]: superfluous space

File: parameter/ParameterMgr.cpp (hunk 3 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 6 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 7 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.h (hunk 9 above)
Comment (clero) [inline parameter/ParameterMgr.h]: done

File: parameter/Results.h (hunk 10 above)
Comment (clero) [inline parameter/Results.h]: done

File: parameter/ParameterMgr.cpp (hunk 4 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 8 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp (hunk 5 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/SystemClass.cpp (hunk 11 above)
Comment (clero) [inline parameter/SystemClass.cpp]: done

**Comment by dawagner** [comment]
rebased onto v2.4.3 and merged in "next".

**Pull request closed without merging**